File: problems/0076-minimum-window-substring/python/solution.py
def minWindow(s, t):
    """Minimum Window Substring — https://leetcode.com/problems/minimum-window-substring/"""
    if not s or not t:
        return ""
    need = {}
    for c in t:
        need[c] = need.get(c, 0) + 1
    required = len(need)
    window = {}
    have = 0
    best_len = float("inf")
    best_left = 0
    left = 0
    for right, c in enumerate(s):
        window[c] = window.get(c, 0) + 1
        if c in need and window[c] == need[c]:
            have += 1
        while have == required:
            if right - left + 1 < best_len:
                best_len = right - left + 1
                best_left = left
            lc = s[left]
            window[lc] -= 1
            if lc in need and window[lc] < need[lc]:
                have -= 1
            left += 1
    return "" if best_len == float("inf") else s[best_left:best_left + best_len]

File: problems/0076-minimum-window-substring/typescript/solution.ts
// Minimum Window Substring — https://leetcode.com/problems/minimum-window-substring/
export function minWindow(s: string, t: string): string {
  if (s.length === 0 || t.length === 0) return "";
  const need = new Map<string, number>();
  for (const c of t) need.set(c, (need.get(c) ?? 0) + 1);
  const required = need.size;
  const window = new Map<string, number>();
  let have = 0;
  let bestLen = Infinity;
  let bestLeft = 0;
  let left = 0;
  for (let right = 0; right < s.length; right++) {
    const c = s[right];
    window.set(c, (window.get(c) ?? 0) + 1);
    if (need.has(c) && window.get(c) === need.get(c)) have++;
    while (have === required) {
      if (right - left + 1 < bestLen) {
        bestLen = right - left + 1;
        bestLeft = left;
      }
      const lc = s[left];
      window.set(lc, (window.get(lc) ?? 0) - 1);
      if (need.has(lc) && (window.get(lc) ?? 0) < (need.get(lc) ?? 0)) have--;
      left++;
    }
  }
  return bestLen === Infinity ? "" : s.slice(bestLeft, bestLeft + bestLen);
}
